text/3857-cfg-version.md

```diff
@@ -0,0 +1,625 @@
+> we can give this condition a semantic name and avoid duplicating it, reducing the chance of bugs:
+> ```rust
+> #[cfg_alias(must_use_exists, since(rust, "1.27"))]
+>
+> #[cfg_attr(must_use_exists, must_use)]
+> fn double(x: i32) -> i32 {
+>     2 * x
+> }
+>
+> fn main() {
+>     double(4);
+>     // warning: unused return value of `double` which must be used
+>     // ^--- This warning only happens if we are on Rust >= 1.27.
+> }
+> ```
+
+Now, let's say `#[cfg(since)]` was stabilized in Rust 1.27 or later, you can check for support for it with:
+```rust
+#[cfg_attr(rust, cfg_attr(since(rust, "1.27"), must_use))]
+fn double(x: i32) -> i32 {
+    2 * x
+}
+
+fn main() {
+    double(4);
+    // warning: unused return value of `double` which must be used
+    // ^--- This warning only happens if we are on Rust >= 1.27.
+}
+```
+However, this would produce an `unexpected_cfgs` lint and you would need to add the following to `Cargo.toml`:
+```toml
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(rust,values(none()))'] }
+```
+
+Say you were wanting to test out `#[must_use]` after it got stabilized on nightly to provide feedback and to be ready for when it hits stable,
+you would instead use `"1.27.0-0"` to match all pre-release versions of 1.27.0:
+```rust
+#[cfg_attr(since(rust, "1.27.0-0"), must_use)]
+fn double(x: i32) -> i32 {
+    2 * x
+}
+
+fn main() {
+    double(4);
+    // warning: unused return value of `double` which must be used
+    // ^--- This warning only happens if we are on Rust >= 1.27.
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `since` cfg predicate
+
+A `since` cfg predicate will be added to Rust.
+As Cargo mirrors Rust's `#[cfg]` syntax, it too will gain this predicate.
+
+The [syntax](https://doc.rust-lang.org/reference/conditional-compilation.html#grammar-ConfigurationPredicate) is:
+```
+ConfigurationVersion -> `since` `(` IDENTIFIER `,` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
+```
+
+If `IDENTIFIER` has no value or is undefined, this will evaluate to `false`.
+
+If `IDENTIFIER` is not a valid [SemVer](https://semver.org/) value, this will evaluate to `false`.
+
+If the string literal does not conform to the syntax from `<major>` to `<major>.<minor>.<patch>-<pre-release>` where the first three fields must be integers, this will evaluate to `false`.
+Note that this excludes support for the `+build` field.
+
+Otherwise, the `IDENTIFIER` will be compared to the string literal according to
+[Cargo's `>=` version requirements](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#comparison-requirements).
+For example, `#[cfg(since(rust, "1.90"))]` will be treated as `1.95.2 >= 1.90.0`.
+
+See also `--check-cfg`.
+
+## `--check-cfg`
+
+A new predicate will be added of the form:
+```
+ConfigurationVersion -> `since` `(` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
+```
```

Comment:
It's unclear to me if we should force a minimum version. We could simply that not providing a minimum supported version means that's it correspond to the lowest SemVer version.